### ISSUE TITLE
Write partitioned parquet: request is not authorized to perform this operation

### ISSUE THREAD

**Issue by UnderTheCraft** [kaaveland/pyarrowfs-adlgen2#17]
Hi, I have an issue while trying to write a dataframe as partitioned parquet. Writing single parquet files works fine, but it seems like the directory creation makes some problems.

The code I am using is the following: 
```python
from azure.identity import CertificateCredential
from azure.storage.filedatalake import DataLakeServiceClient
import pyarrow
import pyarrowfs_adlgen2

credential = CertificateCredential(tenantId, clientId, certificate_data=certificate)
datalake_client = DataLakeServiceClient(account_url=f"https://{accountName}.dfs.core.windows.net", credential=credential)
handler = pyarrowfs_adlgen2.AccountHandler(datalake_service=datalake_client)
fs = handler.to_fs()

# ... create pandas df ... 

path_to_write = f"{container_name}/path/to/partitioned/parquet/base/dir"
df.to_parquet(path_to_write, filesystem=fs, index=False, coerce_timestamps=None, version="2.6", partitioning=["partition_column"]) 

# Alternative approach to write with pyarrow
table = pyarrow.Table.from_pandas(df)
pyarrow.parquet.write_to_dataset(table=table, root_path=path_to_write, filesystem=fs, partitioning=["partition_column"], partitioning_flavor = 'hive')
```

Regardless of which approach I am using, I always get an `HttpResponseError: This request is not authorized to perform this operation.` (see full trace at the bottom).
It seems like the main error comes from the [`create_dir` method](https://github.com/kaaveland/pyarrowfs-adlgen2/blob/77bfac1401055a49790d200753e468b37fd5ab08/pyarrowfs_adlgen2/core.py#L697 ), where a new file_system should be created.
However, in my view, for creating a new directory, `get_file_system_client` needs to be called, passing the `fs_name` (you could also call it container_name I think) as parameter. This returns an `FileSystemClient`, which finally can call the method `create_directory`.
At least this approach worked in my case:
```python
datalake_service_client = DataLakeServiceClient(account_url=f"https://{accountName}.dfs.core.windows.net", credential=credential)
file_service_client = datalake_service_client.get_file_service_client("container_name/fs_name")
file_service_client.create_directory("dir/to/be/created")
```

I am just not sure how this could be implemented or fixed in your library. @kaaveland do you have any idea?

Thanks in advance and best regards
Yannick

---

Complete error trace:

```
/local_disk0/.ephemeral_nfs/envs/pythonEnv-9f78cd2f-a01e-491e-940c-2a933a25b9d4/lib/python3.8/site-packages/pyarrow/dataset.py in write_dataset(data, base_dir, basename_template, format, partitioning, partitioning_flavor, schema, filesystem, file_options, use_threads, max_partitions, max_open_files, max_rows_per_file, min_rows_per_group, max_rows_per_group, file_visitor, existing_data_behavior, create_dir)
    996         scanner = data
    997 
--> 998     _filesystemdataset_write(
    999         scanner, base_dir, basename_template, filesystem, partitioning,
   1000         file_options, max_partitions, file_visitor, existing_data_behavior,

/local_disk0/.ephemeral_nfs/envs/pythonEnv-9f78cd2f-a01e-491e-940c-2a933a25b9d4/lib/python3.8/site-packages/pyarrow/_dataset.pyx in pyarrow._dataset._filesystemdataset_write()

/local_disk0/.ephemeral_nfs/envs/pythonEnv-9f78cd2f-a01e-491e-940c-2a933a25b9d4/lib/python3.8/site-packages/pyarrow/_fs.pyx in pyarrow._fs._cb_create_dir()

/local_disk0/.ephemeral_nfs/envs/pythonEnv-9f78cd2f-a01e-491e-940c-2a933a25b9d4/lib/python3.8/site-packages/pyarrowfs_adlgen2/core.py in create_dir(self, path, recursive)
    695         if recursive or not path:
    696             try:
--> 697                 fs_client = self.create_file_system(fs_name)
    698                 self.file_system_handlers[fs_name] = self.fs_handler_cls(
    699                     fs_client, prefix_fs=True, timeouts=self.timeouts)

/local_disk0/.ephemeral_nfs/envs/pythonEnv-9f78cd2f-a01e-491e-940c-2a933a25b9d4/lib/python3.8/site-packages/pyarrowfs_adlgen2/core.py in create_file_system(self, name)
    661         "datalake_service_timeout")
    662     def create_file_system(self, name):
--> 663         return self.datalake_service.create_file_system(
    664             name, timeout=self.timeouts.datalake_service_timeout
    665         )

/local_disk0/.ephemeral_nfs/envs/pythonEnv-9f78cd2f-a01e-491e-940c-2a933a25b9d4/lib/python3.8/site-packages/azure/storage/filedatalake/_data_lake_service_client.py in create_file_system(self, file_system, metadata, public_access, **kwargs)
    287         """
    288         file_system_client = self.get_file_system_client(file_system)
--> 289         file_system_client.create_file_system(metadata=metadata, public_access=public_access, **kwargs)
    290         return file_system_client
    291 

/local_disk0/.ephemeral_nfs/envs/pythonEnv-9f78cd2f-a01e-491e-940c-2a933a25b9d4/lib/python3.8/site-packages/azure/storage/filedatalake/_file_system_client.py in create_file_system(self, metadata, public_access, **kwargs)
    281         """
    282         encryption_scope_options = kwargs.pop('encryption_scope_options', None)
--> 283         return self._container_client.create_container(metadata=metadata,
    284                                                        public_access=public_access,
    285                                                        container_encryption_scope=encryption_scope_options,

/local_disk0/.ephemeral_nfs/envs/pythonEnv-9f78cd2f-a01e-491e-940c-2a933a25b9d4/lib/python3.8/site-packages/azure/core/tracing/decorator.py in wrapper_use_tracer(*args, **kwargs)
     76             span_impl_type = settings.tracing_implementation()
     77             if span_impl_type is None:
---> 78                 return func(*args, **kwargs)
     79 
     80             # Merge span is parameter is set, but only if no explicit parent are passed

/local_disk0/.ephemeral_nfs/envs/pythonEnv-9f78cd2f-a01e-491e-940c-2a933a25b9d4/lib/python3.8/site-packages/azure/storage/blob/_container_client.py in create_container(self, metadata, public_access, **kwargs)
    318                 **kwargs)
    319         except HttpResponseError as error:
--> 320             process_storage_error(error)
    321 
    322     @distributed_trace

/local_disk0/.ephemeral_nfs/envs/pythonEnv-9f78cd2f-a01e-491e-940c-2a933a25b9d4/lib/python3.8/site-packages/azure/storage/blob/_shared/response_handlers.py in process_storage_error(storage_error)
    187     try:
    188         # `from None` prevents us from double printing the exception (suppresses generated layer error context)
--> 189         exec("raise error from None")   # pylint: disable=exec-used # nosec
    190     except SyntaxError:
    191         raise error

/local_disk0/.ephemeral_nfs/envs/pythonEnv-9f78cd2f-a01e-491e-940c-2a933a25b9d4/lib/python3.8/site-packages/azure/storage/blob/_shared/response_handlers.py in <module>
```

**Comment by kaaveland** [comment]
Thanks for the detailed report! I should be able to make a fix for this + a release in the weekend. A pull request would also be accepted, I don't think it needs to be more than 1 or 2 lines to solve your problem.

What seems to be happening here is that we're using `AccountHandler.create_file_system` in order to retrieve a `FileSystemClient`, but this method has no fallback if we don't have permission to do create containers. The problematic lines are here: https://github.com/kaaveland/pyarrowfs-adlgen2/blob/77bfac1401055a49790d200753e468b37fd5ab08/pyarrowfs_adlgen2/core.py#L697-L699 and here: https://github.com/kaaveland/pyarrowfs-adlgen2/blob/77bfac1401055a49790d200753e468b37fd5ab08/pyarrowfs_adlgen2/core.py#L663-L665

If I recall correctly, `DatalakeServiceClient.create_file_system` is a no-op if the container/file-system already exists, so I wrote it this way with the assumption that it would be simpler than to first check if the file system exists, and create it only if it doesn't, but I hadn't thought of the conflict with RBAC. If you'd like to take a stab at a patch, we should be able to call `DatalakeServiceClient.get_file_service_client` to check whether the container/fs already exists and only issue the create call if it doesn't, but it would probably be good to check that this fails if the file system doesn't exist yet.

**Comment by kaaveland** [comment]
You might also be able to work around the issue by creating a
`FileSystemHandler` in place of `AccountHandler` until I can fix it.

**Comment by kaaveland** [comment]
@UnderTheCraft I am working on this a bit tonight, but I am also encountering other code that makes assumptions about permissions that might not hold for you, depending on how strict your infrastructure is configured. I have a patch ready that fixes the exact case you're reporting here, but there's other stuff you may not be able to do also.

This may not be interesting for you, but there are two "handlers" with the same API that you can pass to `pyarrow`:

- `AccountHandler`, which is convenient when you're working with many containers/file systems. Until now this class had assumed that it has access to create/list file systems, which is what caused your problem. This one essentially treats a container/filesystem as a directory that is a child to the root directory, which is the storage account.
- `FileSystemHandler`, which is scoped to only one container/filesystem. This one doesn't "know" about the storage account, and it's used internally by `AccountHandler` (we create 1 for each container/filesystem that we access).

If your credential does not have access to list containers/file systems, I fear you'll encounter a new problem by trying to use `AccountHandler`. I would need to change a fair bit of internal logic to remove the assumption that we can list file systems, so I'm not sure when I would be able to make a patch for that. If that happens, you should try to see if you can use `FileSystemHandler`, you other than needing an extra argument in the constructor or `from_account_name`, it has the same API. I'll try to push https://github.com/kaaveland/pyarrowfs-adlgen2/pull/18 to pypi tomorrow.

**Comment by kaaveland** [comment]
Released 0.2.4 at https://pypi.org/project/pyarrowfs-adlgen2/0.2.4/ -- closing ticket for now, feel free to reopen/report again if you encounter another problem.

**Comment by UnderTheCraft** [comment]
Thanks a lot @kaaveland for your effort!
So far, read/write seems to work perfectly fine (using AccountHandler as well as FileSystemHandler). That fix really helped me a lot. :)
